iroh/src/magicsock/transports/relay.rs

```diff
@@ -0,0 +1,423 @@
+use crate::magicsock::RelayContents;
+
+mod actor;
+
+pub(crate) use self::actor::Config as RelayActorConfig;
+use self::actor::{RelayActor, RelayActorMessage, RelayRecvDatagram, RelaySendItem};
+
+#[derive(Debug)]
+pub(crate) struct RelayTransport {
+    /// Queue to receive datagrams from relays for [`quinn::AsyncUdpSocket::poll_recv`].
+    relay_datagram_recv_queue: mpsc::Receiver<RelayRecvDatagram>,
+    /// Channel on which to send datagrams via a relay server.
+    relay_datagram_send_channel: mpsc::Sender<RelaySendItem>,
+    actor_sender: mpsc::Sender<RelayActorMessage>,
+    _actor_handle: AbortOnDropHandle<()>,
+    my_relay: Watchable<Option<RelayUrl>>,
+    my_node_id: NodeId,
+}
+
+impl RelayTransport {
+    pub(crate) fn new(config: RelayActorConfig) -> Self {
+        let (relay_datagram_send_tx, relay_datagram_send_rx) = mpsc::channel(256);
+
+        let (relay_datagram_recv_tx, relay_datagram_recv_rx) = mpsc::channel(512);
+
+        let (actor_sender, actor_receiver) = mpsc::channel(256);
+
+        let my_node_id = config.secret_key.public();
+        let my_relay = config.my_relay.clone();
+
+        let relay_actor = RelayActor::new(config, relay_datagram_recv_tx);
+
+        let actor_handle = AbortOnDropHandle::new(task::spawn(
+            async move {
+                relay_actor
+                    .run(actor_receiver, relay_datagram_send_rx)
+                    .await;
+            }
+            .instrument(info_span!("relay-actor")),
+        ));
+
+        Self {
+            relay_datagram_recv_queue: relay_datagram_recv_rx,
+            relay_datagram_send_channel: relay_datagram_send_tx,
+            actor_sender,
+            _actor_handle: actor_handle,
+            my_relay,
+            my_node_id,
+        }
+    }
+
+    pub(crate) fn create_sender(&self) -> RelaySender {
+        RelaySender {
+            sender: PollSender::new(self.relay_datagram_send_channel.clone()),
+        }
+    }
+
+    pub(super) fn poll_recv(
+        &mut self,
+        cx: &mut Context,
+        bufs: &mut [io::IoSliceMut<'_>],
+        metas: &mut [quinn_udp::RecvMeta],
+        source_addrs: &mut [Addr],
+    ) -> Poll<io::Result<usize>> {
+        let mut num_msgs = 0;
+        for ((buf_out, meta_out), addr) in bufs
+            .iter_mut()
+            .zip(metas.iter_mut())
+            .zip(source_addrs.iter_mut())
+        {
+            let dm = match self.relay_datagram_recv_queue.poll_recv(cx) {
+                Poll::Ready(Some(recv)) => recv,
+                Poll::Ready(None) => {
+                    error!("relay_recv_channel closed");
+                    return Poll::Ready(Err(io::Error::new(
+                        io::ErrorKind::NotConnected,
+                        "connection closed",
+                    )));
+                }
+                Poll::Pending => {
+                    break;
+                }
+            };
+
+            buf_out[..dm.buf.len()].copy_from_slice(&dm.buf);
```

Comment:
Long, long in the future I can now see us polling the TCP socket from the relay server directly during this poll, allowing us to avoid this copy.  Especially once we don't have any non-quic messages left since then only very few times we'll need to send a non-QUIC message over the relay protocol that should not be passed up.